### PR TITLE
Use babel, eslint

### LINKED ISSUE #116: Use babel for transpilation
**Is your feature request related to a problem? Please describe.**
Typescript is good at type checking, but not as good at transpilation to various JS targets. The way it is configured leads to some limitations regarding which javascript features we can use.

Because we want to support a wide variety of javascript targets (eg: the edge browser, old versions of node, etc, etc), we will rightfully (and currently do!) maintain a conservative `target` in our `tsconfig.json`. This `target` option sets how the typescript is transpiled to vanilla javascript, but unfortunately, it also opens/closes various typescript features. So we can't use the full typescript feature set without also excluding many older targets! And if we use an older target, now we must hurt our devEx/productivity by reverting to older, invariably more verbose coding standards even as alternatives exist.

It doesn't have to be this way. We can have our cake and eat it too. (Ie use the latest-and-greatest javascript and typescript features and also target any version of javascript we want)
Since we have to transpile our TS code to vanilla JS one way or another, we may as well use the most flexible transpiler available. Especially if that option doesn't remove or conflict with the type-checking that typescript provides.

[Babel](https://babeljs.io/) is this transpiler. It makes it easy to target various versions of JS. It is widely used and supported in the wider javascript community. It is compatible with typescript. It can easily be configured to inject any necessary polyfills, rewrite unsupported syntax, etc. just as typescript does, but better, more configurably, more featurefully, more quickly, more flexibly.
Using it along with typescript gives us the best of both worlds.

one example:
current situation: want to use native BigInts but also target the Edge browser? too bad, can't even contemplate
future situation: want to use native BigInts but also target the Edge browser? ezpz, we can also make a build optimized for node, the latest chrome, etc, etc.

**Describe the solution you'd like**
Set the `target` in `tsconfig.json` to `esnext`, for the full typescript feature set
Use typescript to check types (in IDE and through CLI/CI)
Use babel features that don't conflict with typescript
Use babel to compile builds (create at minimum, one build for node, one build for "browsers")

**Describe alternatives you've considered**
- Stick with using typescript for transpilation and:
  - target es5 (sad, we can't use BigInts or other productivity-enhancing features)
  - target esnext (sad, if we use certain features, we can't run our code on IE8, edge, safari, etc etc)
- Use babel transpilation and
  - w/o taking advantage of any esnext features (sad, but still better than current situation)
  - with typescript-incompatible features (sad, we want to retain typescript compatibility)

**Additional context**
https://iamturns.com/typescript-babel/
https://github.com/Microsoft/TypeScript-Babel-Starter

Writing code in a backward-compatible way hurts dev-ex and productivity and is largely unnecessary. Lets use our language's full feature set (or at least position ourselves to take this advantage if we so desire in the future) and let the computer take care of the backward-compatible thing on its own, when it needs to do so.

### PR DESCRIPTION
**Pull request by wemeetagain**
~Relies on #120 , once that lands I can rebase off master~
Resolves #116 

- Use eslint instead of tslint
- Use babel in mocha tests
- Use esnext tsconfig target
- Add check-types script to do typescript checking via cli
- Add pretest type checking
- eslint uncovered significantly more linting errors than tslint, fixed those

### CODE REVIEW

File: beaconChain/src/helpers/stateTransitionHelpers.ts [hunk 1]
```diff
@@ -19,27 +23,37 @@ import {
...
   blsAggregatePubkeys,
   blsVerifyMultiple,
-} from "../lib/stubs/bls";
+  blsVerify,
+} from "../stubs/bls";
 
-import { hashTreeRoot } from "../lib/state";
+
+// This function was copied from ssz-js
+// TODO: either export hash from ssz-js or move to a util-crypto library
```
Comment (GregTheGreek) [inline beaconChain/src/helpers/stateTransitionHelpers.ts:53]: Let's export from SSZ-JS and cut a new release in a week

File: beaconChain/src/helpers/stateTransitionHelpers.ts [hunk 2]
```diff
@@ -119,7 +133,7 @@ export function getActiveValidatorIndices(validators: Validator[], epoch: Epoch)
 function shuffle<T>(values: T[], seed: bytes32): T[] {
   const valuesCount: int = values.length;
   // Entropy is consumed from the seed in 3-byte (24 bit) chunks.
-  const randBytes: number = 3;
+  const randBytes = 3;
```
Comment (GregTheGreek) [inline beaconChain/src/helpers/stateTransitionHelpers.ts:136]: why did you remove `: number`

File: beaconChain/src/helpers/stateTransitionHelpers.ts [hunk 3]
```diff
@@ -130,7 +144,7 @@ function shuffle<T>(values: T[], seed: bytes32): T[] {
   // Make a copy of the values
   const output: T[] = values.slice();
   let source: bytes32 = seed;
-  let index: number = 0;
+  let index = 0;
```
Comment (GregTheGreek) [inline beaconChain/src/helpers/stateTransitionHelpers.ts:147]: again?

**Review by GregTheGreek — CHANGES_REQUESTED** [review]
Just a few little things

File: beaconChain/src/helpers/stateTransitionHelpers.ts (hunk 1 above)
Comment (wemeetagain) [reply beaconChain/src/helpers/stateTransitionHelpers.ts:53]: The block processing PR already imports from ssz-js. I figured we may as well use it if we got it, rather than importing and using two different places, the stub and the real thing.

File: beaconChain/src/helpers/stateTransitionHelpers.ts (hunk 2 above)
Comment (wemeetagain) [reply beaconChain/src/helpers/stateTransitionHelpers.ts:136]: This was removed by a eslint --fix.
I think theres a rule to the effect of: no super obvious types.
something like `const b: boolean = true;` would likewise get flagged. (see the isPowerOfTwo function where that happened and was changed)

Comment (GregTheGreek) [reply beaconChain/src/helpers/stateTransitionHelpers.ts:136]: That actually makes sense...

File: beaconChain/src/helpers/stateTransitionHelpers.ts (hunk 1 above)
Comment (GregTheGreek) [reply beaconChain/src/helpers/stateTransitionHelpers.ts:53]: Makes sense.

**Comment by GregTheGreek** [comment]
The CI is failing on this PR definitely switched tabs by accident.

**Review by GregTheGreek — APPROVED** [review]

**Review by ansermino — APPROVED** [review]

**Comment by GregTheGreek** [comment]
@wemeetagain Please rebase :)

**Comment by wemeetagain** [comment]
Rebased off latest master (#120 merged)
